MigrationV3V4/Node/V4V3-user-state-adapter-sample-bot/V4V3StorageMapper/src/index.ts

```diff
@@ -0,0 +1,158 @@
+/**
+ * @module v4-v3-storage-mapper
+ */
+/**
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { Storage, StoreItems } from 'botbuilder';
+import * as validate from 'uuid-validate';
+
+export class StorageMapper implements Storage {
+
+  private storageType: string;
+  private v3StorageProvider: any;
+  private userId: string;
+  
+  public constructor(v3StorageProvider: any) {
+    if (!v3StorageProvider) {
+      throw new Error('A storage provider must be provided.');
+    }
+
+    this.storageType = v3StorageProvider.storageClient.constructor.name;
+    this.v3StorageProvider = v3StorageProvider;
+    this.userId = '';
+  }
+
+  private extractUserId(key: string): string {
+    const keySegments: Array<string> = key.split(',');
+    const user_id: string = keySegments.find(segment => {
+      return validate(segment);
+    });
+    this.userId = user_id;
+    return user_id;
+  }
+
+  private formatDataResponse = (key, inputdata) => {
+    const itemKey = `${key ? key : this.userId},userData`;
+    const formattedData = {};
+    formattedData[itemKey] = {};
+    formattedData[itemKey]['userProfile'] = inputdata;
+    return formattedData;
+  };
+  
+  public read(keys: string[]): Promise<StoreItems> {
+    if (!keys || keys.length === 0) {
+      return Promise.resolve({});
+    }
+
+    const userStateKey: string = keys.find(key => {
+      return key.includes("userData");
+    });
+
+    if (userStateKey) {
+      const context = {
+        userId: this.extractUserId(userStateKey),
+        persistUserData: true,
+        persistConversationData: false
+      };
+    
+      switch (this.storageType) {
```

Comment:
I'm not sure why this switch is here ... if all of the v3StorageProviders implement the same interface, can't we just call the same method for each of these?

Reply:
The v3 storage providers don't have a concept of interfaces or classes - they use vanilla JS prototypes. The method signatures and arguments are different across the three are different so that' why I needed a switch.

Reply:
Update: no switch needed. While there's no interface there is a wrapper component that interfaces with the 3 providers which I wasn't using. Code updated to use this!